File: 0002-add-two-numbers.py
"""
2. Add Two Numbers

Submitted: October 21, 2025

Runtime: 6 ms (beats 32.71%)
Memory: 18.20 MB (beats 45.96%)
"""

# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

def list_iter(p: ListNode):
    while p:
        yield p.val
        p = p.next

def iter_to_number(it):
    it = iter(it)
    n = 0
    for digit in it:
        n *= 10
        n += digit
    return n

class Solution:
    def addTwoNumbers(self, l1: Optional[ListNode], l2: Optional[ListNode]) -> Optional[ListNode]:
        ll1 = tuple(list_iter(l1))[::-1]
        ll2 = tuple(list_iter(l2))[::-1]
        n1 = iter_to_number(ll1)
        n2 = iter_to_number(ll2)
        n3 = n1 + n2
        resl = []
        if n3 == 0:
            return ListNode()
        else:
            while n3 != 0:
                resl.append(n3 % 10)
                n3 //= 10
        head = ListNode(resl[0])
        prev = head
        tail: ListNode
        for n in resl[1:]:
            tail = ListNode(n)
            prev.next = tail
            prev = tail
        return head

File: 0002-add-two-numbers.cpp
/**
2. Add Two Numbers

Submitted: Feb 01, 2026

Runtime: 6 ms (beats 8.82%)
Memory: 77.92 MB (beats 77.92%)
*/

/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* addTwoNumbers(ListNode* l1, ListNode* l2) {
        int carry = 0;
        vector<int> v1, v2, v3;
        
        for (auto p = l1; p != nullptr; p = p->next) {
            v1.push_back(p->val);
        }
        for (auto p = l2; p != nullptr; p = p->next) {
            v2.push_back(p->val);
        }

        int n = v1.size(), m = v2.size();
        int digits = max(n, m);
        v1.resize(digits);
        v2.resize(digits);

        for (int i = 0; i < digits; ++i) {
            // cout << v1[i] << ' ' << v2[i] << '\n';
            int add = v1[i] + v2[i] + carry;
            int digit = add % 10;
            carry = add / 10;
            // cout << add << ' ' << digit << ' ' << carry << '\n';
            v3.push_back(digit);
        }

        ListNode* res = digits == n ? l1 : l2;
        {
            ListNode* p = digits == n ? l1 : l2;
            for (int i = 0; i < digits; ++i) {
                // cout << p << ' ' << p->next << ' ';
                p->val = v3[i];
                p = p->next;
            }
        }

        if (carry > 0) {
            // v3.push_back(carry);
            ListNode* end;
            ListNode* p = digits == n ? l1 : l2;
            while (p->next != nullptr) {
                p = p->next;
            }
            end = p;
            p = new ListNode(carry);
            end->next = p;
        }
        
        return res;
    }
};